Fix malformed ActivityPub handles for email-based logins

obenland opened the pull request:
Fixes #2070.

## Proposed changes:
* Modified `get_preferred_username()` method in `includes/model/class-user.php` to detect email-based logins and sanitize them using `sanitize_title()`
* Added comprehensive test coverage in `tests/includes/model/class-test-user.php` to verify proper email username sanitization
* Ensures ActivityPub handles are properly formatted as `@username@domain.com` instead of malformed `@username@gmail.com@domain.com`

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:

### Setup
1. Install and activate the ActivityPub plugin in a WordPress test environment
2. Install and configure Site Kit with Google authentication (or manually create a user with email-based login)

### Test Case 1: Email-based Login from Site Kit
1. Create a test user via Site Kit Google login, which typically creates a username like `kimjiwoon75@gmail.com`
2. Navigate to the user's ActivityPub profile page or use the webfinger endpoint
3. **Expected result**: ActivityPub handle should be `@kimjiwoon75-gmail-com@yourdomain.com`
4. **Previous buggy behavior**: Handle was malformed as `@kimjiwoon75@gmail.com@yourdomain.com`

### Test Case 2: Manual Email-based Username
1. Create a user with login name `testuser@example.com`
2. Check their ActivityPub handle via webfinger or profile
3. **Expected result**: Handle should be `@testuser-example-com@yourdomain.com`

### Test Case 3: Normal Username (Regression Test)
1. Create a user with normal login name `normaluser`
2. Check their ActivityPub handle
3. **Expected result**: Handle should be `@normaluser@yourdomain.com` (unchanged behavior)

### Test Case 4: Run Unit Tests
```bash
npm run env-test -- --filter=test_email_username_sanitization
```
**Expected result**: All tests should pass

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch

#### Type

-   [x] Fixed - for any bug fixes

#### Message

Fix malformed ActivityPub handles for users with email-based logins (e.g., from Site Kit Google authentication)

</details>

obenland committed:
Fix malformed ActivityPub handles for email-based logins

Sanitize email-based usernames (e.g., from Site Kit Google login) to prevent
malformed ActivityPub handles like @username@gmail.com@domain.com.

- Modified get_preferred_username() to detect and sanitize email logins
- Added comprehensive test coverage for email username sanitization
- Ensures proper webfinger format without double @ symbols

Fixes #2070.

matticbot committed:
Add changelog

obenland committed:
Add global namespace to str_contains in User model

Prefixed str_contains with a backslash to ensure the global PHP function is used, preventing potential issues with overridden or namespaced functions.

obenland committed:
Merge branch 'fix/email-username-sanitization' of https://github.com/Automattic/wordpress-activitypub into fix/email-username-sanitization

obenland committed:
Fix code style alignment in test

Fix PHPCS alignment warnings for variable assignments.

Jiwoon-Kim commented:
If the administrator creates a new account using a Google email address but sets the username in a non-email format (a regular ID instead of an email), and then links it via Sign in with Google, this bug can be bypassed.

Honestly, I’m not entirely sure whether this should be debugged on the ActivityPub side, or if it’s something that Site Kit should address by changing how it handles email-based username creation.

If my memory serves me correctly, Misskey prevents users from creating accounts with email-style IDs at signup (or at least explicitly instructs users not to enter them).

---

However, since the author page ID for email-based accounts does not include special characters, this fix still seems to have some value.

Jiwoon-Kim commented:
The profile URL is generated as:
https://travel-in-busan.com/author/kimjiwoon75gmail-com/

To maintain consistency with the ActivityPub handle, instead of replacing @ with a hyphen, the @ should be omitted entirely.

👉 Desired format:
@kimjiwoon75gmail-com@domain.com

---
is not very “clean” as a username.

It may be worth considering a redirect mechanism in the future so that such legacy usernames can point to a cleaner, newly chosen identifier.

Of course, this would also require solving the name collision problem (e.g., if two users try to adopt the same “cleaned” username).

👉 In practice, this would mean:

Keep the legacy handle for backwards compatibility.

Allow the user/admin to register a new canonical handle.

Set up a redirect/alias system so that old mentions and followers still resolve correctly.

---

This should be addressed together with the ability for users to edit their blog profile handle. For example, mechanisms that allow arbitrary manipulation of the handle—such as using the posts page slug as the handle—should be prevented. The goal is to maintain consistency and integrity of ActivityPub handles across posts and profiles.

obenland committed:
Fix email username sanitization logic

Use explicit string replacement instead of sanitize_title()
to ensure dots are converted to dashes as expected by tests.
Also use filter_var() for proper email validation.

Jiwoon-Kim commented:
https://wordpress.org/support/topic/activitypub-handle-format-issue-with-google-login/#post-18612844

pfefferle commented:
@obenland you use a different formatter than core, so it is not possible to run a webfinger lookup. I think core simply removes the `@` instead of replacing it with a `-`.

pfefferle commented on includes/model/class-user.php:
```diff
@@ -169,7 +169,14 @@ public function get_alternate_url() {
+
+		// Handle cases where login is an email address (e.g., from Site Kit Google login).
+		if ( \filter_var( $login, FILTER_VALIDATE_EMAIL ) ) {
+			$login = \str_replace( array( '@', '.' ), '-', $login );
```
What about:

```suggestion
			$login = \sanitize_title( $login );
```

???

pfefferle commented on includes/model/class-user.php:
```diff
@@ -169,7 +169,14 @@ public function get_alternate_url() {
+
+		// Handle cases where login is an email address (e.g., from Site Kit Google login).
+		if ( \filter_var( $login, FILTER_VALIDATE_EMAIL ) ) {
+			$login = \str_replace( array( '@', '.' ), '-', $login );
```
this might be even better and should be the same result:

```suggestion
			$login = \get_the_author_meta( 'user_nicename', $this->_id );
```

obenland replied:
I guess it comes down to the expectations we have towards what makes for an appealing handle. a49da59021249d38848cd043302e50943fa763a3 moved away from `sanitize_title()` to transition fediverse handles from `usernamegmail-com` to `username-gmail-com`.

The latter might be a bit nicer, but happy to defer if you have another preference. I don't feel strongly about it.

pfefferle replied:
It might look nicer, but it breaks the lookup! If you stick with the custom formatter you would have to also add support to the Users::get_by_* functions

obenland replied:
Thanks that's helpful!

obenland committed:
Use user_nicename for preferred username from email

Updated the logic in User::get_preferred_username() to use the user's nicename when the login is an email address, instead of replacing '@' and '.' with dashes. Adjusted related tests to match the new behavior.

obenland committed:
Merge branch 'trunk' into fix/email-username-sanitization
